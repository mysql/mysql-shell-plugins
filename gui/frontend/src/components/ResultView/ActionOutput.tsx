/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

import { Monaco } from "../ui/CodeEditor";
import { Label, Component, IComponentProperties, Orientation, Container } from "../ui";
import { ResultStatus } from ".";
import { CodeEditor } from "../ui/CodeEditor/CodeEditor";
import { ResultTextLanguage } from "./";
import { IExecutionInfo } from "../../app-logic/Types";

interface IActionOutputProperties extends IComponentProperties {
    text: string;
    language?: ResultTextLanguage;

    executionInfo?: IExecutionInfo; // The status message to show. Can also be an error messages.
}

// A component to render a given text with syntax highlighting (depending on the given language).
export class ActionOutput extends Component<IActionOutputProperties> {

    private labelRef = React.createRef<HTMLLabelElement>();

    public constructor(props: IActionOutputProperties) {
        super(props);

        this.addHandledProperties("text", "language", "isError", "executionInfo");
    }

    public componentDidMount(): void {
        this.colorize();
    }

    public componentDidUpdate(): void {
        this.colorize();
    }

    public render(): React.ReactNode {
        const { text, language, executionInfo } = this.props;
        const className = this.getEffectiveClassNames(["resultText"]);

        return (
            <Container
                orientation={Orientation.TopDown}
                className={className}
            >
                <Label
                    innerRef={this.labelRef}
                    data-lang={language}
                    caption={text}
                    data-tooltip=""
                />
                {executionInfo && <ResultStatus executionInfo={executionInfo} />}

            </Container>
        );
    }

    /**
     * Uses Monaco to convert text into colorized tokens, depending on the language property.
     */
    private colorize(): void {
        /* istanbul ignore next */
        if (!this.labelRef?.current) {
            return;
        }

        const { language } = this.props;

        if (language && language !== "ansi") {
            void Monaco.colorizeElement(this.labelRef?.current, {
                theme: CodeEditor.currentThemeId,
            });
        }
    }
}
