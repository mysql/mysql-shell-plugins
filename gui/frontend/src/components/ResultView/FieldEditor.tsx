/*
 * Copyright (c) 2023, 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { Buffer } from "buffer";

import { Image } from "../ui/Image/Image.js";

import { DBDataType, DialogResponseClosure, IDialogRequest } from "../../app-logic/Types.js";
import { AwaitableValueEditDialog } from "../Dialogs/AwaitableValueEditDialog.js";
import { IDialogSection, IDialogValidations, IDialogValues } from "../Dialogs/ValueEditDialog.js";
import { HexEditor, HexValueGrouping } from "../ui/HexEditor/HexEditor.js";

export interface IFieldEditorData extends IDictionary {
    /** The type of the data, expressed as database column data type. */
    dataType: DBDataType;

    /** Base64 encoded binary data. */
    value: string;
}

export class FieldEditor extends AwaitableValueEditDialog {
    static #typesWithImageData = new Set<DBDataType>([
        DBDataType.Blob,
        DBDataType.MediumBlob,
        DBDataType.LongBlob,
        DBDataType.Binary,
        DBDataType.Varbinary,
    ]);

    protected get id(): string {
        return "fieldEditor";
    }

    public override async show(request: IDialogRequest): Promise<IFieldEditorData | DialogResponseClosure> {
        const data = request.parameters as IFieldEditorData;

        const dialogValues = this.dialogValues(request, data.dataType, data.value);
        const result = await this.doShow(() => {
            return dialogValues;
        }, { title: "Database Field Editor" });

        if (result.closure === DialogResponseClosure.Accept) {
            return {
                ...this.processResults(result.values),
                dataType: data.dataType,
            } as IFieldEditorData;
        }

        return DialogResponseClosure.Cancel;
    }

    protected override validateInput = (closing: boolean, values: IDialogValues): IDialogValidations => {
        const result: IDialogValidations = {
            messages: {},
            requiredContexts: [],
        };

        const mainSection = values.sections.get("mainSection");

        if (closing) {
            if (mainSection) {
                //
            }
        } else if (mainSection) {
            //
        }

        return result;
    };

    private dialogValues(request: IDialogRequest, dataType: DBDataType, data?: string | null): IDialogValues {
        // Content is stored as base64 encoded binary data. We need to decode it.
        let content = "";
        if (data) {
            const buffer = Buffer.from(data, "base64");
            content = buffer.toString();
        }

        const firstPart = content.substring(0, 10).toLowerCase();
        let imagePrefix = "";
        if (firstPart.indexOf("png") > -1) {
            imagePrefix = "data:image/png;base64,";
        } else if (firstPart.indexOf("jpg") > -1) {
            imagePrefix = "data:image/jpg;base64,";
        }

        const textSection: IDialogSection = {
            caption: "Text",
            groupName: "editorSections",
            expand: true,
            values: {
                value: {
                    type: "text",
                    caption: "Value",
                    multiLine: true,
                    value: content,
                    horizontalSpan: 8,
                },
            },
        };

        // Convert the content to a byte array.
        const encoder = new TextEncoder();
        const bytes = encoder.encode(content);
        const hexSection: IDialogSection = {
            caption: "Binary",
            groupName: "editorSections",
            expand: true,
            values: {
                value: {
                    type: "custom",
                    horizontalSpan: 8,
                    component: <HexEditor value={bytes.buffer} grouping={HexValueGrouping.Byte} />,
                },
            },
        };

        const imageSection: IDialogSection = {
            caption: "Image",
            groupName: "editorSections",
            expand: true,
            values: {
                value: {
                    type: "custom",
                    horizontalSpan: 8,
                    component: <Image src={`${imagePrefix}${data ?? ""}`} />,
                },
            },
        };

        const jsonSection: IDialogSection = {
            caption: "JSON",
            groupName: "editorSections",
            expand: true,
            values: {
                value: {
                    type: "text",
                    caption: "Value",
                    multiLine: true,
                    value: content,
                    horizontalSpan: 8,
                },
            },
        };

        const sections = new Map<string, IDialogSection>([
            ["textSection", textSection],
            ["hexSection", hexSection],
        ]);

        if (FieldEditor.#typesWithImageData.has(dataType) && imagePrefix.length > 0) {
            sections.set("imageSection", imageSection);
        }

        if (dataType === DBDataType.Json) {
            sections.set("jsonSection", jsonSection);
        }

        return {
            id: "mainSection",
            sections,
        };
    }

    private processResults = (dialogValues: IDialogValues): IDictionary => {
        const mainSection = dialogValues.sections.get("mainSection");

        if (mainSection) {
            const values = {
                dataType: mainSection.values.dataType.value as DBDataType,
                value: mainSection.values.value.value as string,
            };

            return values;
        }

        return {};
    };

}
