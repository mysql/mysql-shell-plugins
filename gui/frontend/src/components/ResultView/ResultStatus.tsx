/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";
import { isNil } from "lodash";

import { Component, Label, Container, Orientation, IComponentProperties, Message } from "../ui";
import { IExecutionInfo, MessageType } from "../../app-logic/Types";

export interface IResultStatusProperties extends IComponentProperties {
    executionInfo: IExecutionInfo;
}

// Implements a text output area usually used for execution results (except DB data).
export class ResultStatus extends Component<IResultStatusProperties> {

    public constructor(props: IResultStatusProperties) {
        super(props);

        this.addHandledProperties("executionInfo");
    }

    public render(): React.ReactNode {
        const { executionInfo, children } = this.props;

        let text;
        let messageClass = "";
        if (!isNil(executionInfo.type) && executionInfo.type !== MessageType.Response) {
            messageClass = "containsMessage";
            text = <Message type={executionInfo.type}>{executionInfo.text}</Message>;
        } else {
            text = <Label>{executionInfo.text}</Label>;
        }

        const className = this.getEffectiveClassNames(["resultStatus", messageClass]);

        return (
            <Container
                className={className}
                orientation={Orientation.LeftToRight}
                {...this.unhandledProperties}
            >
                {text}
                {isNil(executionInfo.type) && children}
            </Container>
        );
    }

}
