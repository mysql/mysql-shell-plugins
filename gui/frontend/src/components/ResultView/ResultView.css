/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

/* Temporarily hide apply/revert buttons */
.msg.resultHost #editSeparator,
.msg.resultHost #applyButton,
.msg.resultHost #revertButton {
    display: none;
}

.msg.resultHost {
    flex: 1 1 auto;
    border-radius: 8px;
}

.msg.textHost {
    flex: 1 1 auto;
}

.msg.resultHost:not(.dropTarget) > .content {
    background: transparent;
}

.msg.codeEditor .zoneHost .renderTarget > .divider {
    cursor: ns-resize;
}

.msg.codeEditor .zoneHost .renderTarget > .divider.minimum {
    cursor: s-resize;
}

.msg.codeEditor .zoneHost .renderTarget > .divider.maximum {
    cursor: n-resize;
}

.msg.resultHost .resultTabview .tabArea {
    background: transparent;
}

.msg.resultHost {
    font-family: "SourceCodePro+Powerline+Awesome+MySQL";
    height: 100%;

    align-content: stretch;
}

.msg.resultHost .resultTabview .outputHost {
    flex: 1 1 auto;
    background: var(--editorZone-background);
}

.msg.resultHost .resultTabview .outputHost .cmdIndex {
    color: var(--resultStatus-foreground);
    margin-right: 4px;
}

.msg.resultHost .resultTabview .outputHost > .labelHost {
    min-height: 1.3em;
}

.msg.resultHost .resultView .tabulator {
    flex: 1 1 auto;
}

.msg.resultHost .resultView .tabulator-tableholder {
    background-color: var(--resultView-background);
    color: var(--resultView-foreground);

    /* Scrollbar appearance for Firefox. Other browser use the --webkit-scrollbar pseudo elements. */
    scrollbar-width: auto;
    scrollbar-color: var(--scrollbarSlider-background) var(--scrollbarSlider-shadow);
}

.msg.resultHost .resultView .tabulator .tabulator-header {
    background-color: var(--resultView-headerBackground);
    color: var(--resultView-headerForeground);
    font-weight: normal;

    border: none;
}

.msg.resultHost .resultView .tabulator .tabulator-header .tabulator-arrow {
    border-bottom-color: var(--resultView-headerForeground) !important;
}

.msg.resultHost .resultView .tabulator .tabulator-header .tabulator-col {
    background: transparent;
    border-right-color: var(--resultView-gridColor);
}

.msg.resultHost .resultView .tabulator .tabulator-header .tabulator-col .tabulator-col-title {
    padding: 0 0;
}

.msg.resultHost .resultView .tabulator .tabulator-row {
    padding: 0;
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell {
    padding: 2px 9px 2px 10px;

    /* These properties ensure ellipses on large multi line content */
    display: -webkit-inline-box;
    -webkit-line-clamp: 4;
    -webkit-box-orient: vertical;

    overflow: hidden;
    white-space: pre-wrap;
}

.msg.resultHost .resultView .tabulator.verticalGrid .tabulator-row .tabulator-cell {
    border-right: 1px solid var(--resultView-gridColor);
}

.msg.resultHost .resultView .tabulator.horizontalGrid .tabulator-row .tabulator-cell {
    border-bottom: 1px solid var(--resultView-gridColor);
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .iconHost {
    height: 14px;
    padding-top: 4px;
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .iconHost > .icon {
    background-color: var(--resultView-foreground);
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .cellEditorHost {
    margin: 0 0 -4px -2px;
    width: 100%;
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .cellEditorHost .input {
    resize: none;
    padding: 0 2px;

    max-height: 100px;
    width: 100%;

    border: none;
    font: inherit;
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .labelHost {
    width: 100%;
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .labelHost .label {
    white-space: pre-line;
    max-height: 100px;
    overflow: scroll;
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .checkBoxHost {
    height: 18px;
}

.msg.resultHost .resultView .tabulator .tabulator-row .tabulator-cell .checkBoxHost > .checkbox {
    height: 18px;
}

/* A message which is part of a result view. */
.msg.resultHost .resultStatus .message {
    padding: 4px 8px;
    border: none;
    border-radius: 0;
    flex: 1 1 auto;
}

/* Messages embedded in a result status have a different background, so they must fill the entire status. */
.msg.resultHost .resultStatus.containsMessage {
    padding-left: 0;
    padding-right: 0;
}

.msg.resultHost .resultStatus {
    color: var(--resultStatus-foreground);
    background-color: var(--resultStatus-background);

    font-size: 0.8rem;
    min-height: 28px;

    padding: 0 4px;
    overflow: hidden;
}

.msg.resultHost .resultView .label {
    white-space: pre-wrap;
}

.zoneHost .resultText {
    padding: 0 8px;

    background-color: var(--editorZone-background);
    color: var(--resultZone-foreground);
    overflow: auto;
    white-space: pre-wrap;
}

.msg.resultHost .resultText > .label {
    overflow: auto;
    flex: 1 1 auto;
}

.msg.resultHost .resultStatus .toolbar {
    flex: 0 0 auto;
    min-width: fit-content;

    padding: 2px 0;

    border: none;
    background: none;

    color: var(--resultStatus-foreground);
}

.msg.resultHost .resultStatus > .label {
    flex: 1 1 auto;

    margin: 2px 4px 2px 4px;
    white-space: normal;
    align-self: center;
}

.msg.resultHost .resultStatus .toolbar .button {
    padding: 0;
    margin: 0 6px;

    background: none;

    align-items: center;

    border-radius: 2px;
    width: 20px;
    height: 20px;
}

.msg.resultHost .resultStatus .toolbar .button > .icon {
    width: 18px;
    height: 18px;

    background-color: var(--resultStatus-foreground);
}

.msg.resultHost .resultStatus .toolbar .divider {
    height: 20px;
    margin: 0 5px;
}

.msg.resultHost .resultStatus .toolbar .dropdown .icon {
    flex: 0 0 24px;
    width: 24px;
    height: 18px;

    margin: 0;
    background-color: var(--resultStatus-foreground);
}

.msg.resultHost .resultStatus .toolbar #viewStyleDropDown {
    padding: 0;
    margin: 0 9px 0 6px;
    flex: 0 0 40px;

    min-width: 0;
    min-height: 0;
}

.msg .renderTarget .yLabel {
    font-size: 3em;
    transform: translate(50%, 50px)
}

.msg .renderTarget polyline {
    opacity: .75;
    stroke: var(--editorZone-foreground);
    stroke-width: 1px;
    fill: none;
}

.msg.resultHost .resultView .tabulator .tabulator-placeholder {
    background-color: var(--resultView-background);
}

.msg.codeEditor .zoneMargin {
    margin-left: 35px;
    background-color: var(--editorZone-background);
}

.msg.codeEditor .zoneMargin {
    display: block;
    width: 100%;
    height: inherit;
}

.msg.codeEditor .zoneMargin::before {
    content: "";
    display: block;
    position: absolute;

    width: 7px;
    right: 45px;
    top: 0;
    height: 100%;

    mask-image: url(../../assets/images/editor-side-line-straight.svg);
    background-color: var(--editorPrompt-result);
    mask-size: 5px;
    mask-repeat: repeat-y;
}
