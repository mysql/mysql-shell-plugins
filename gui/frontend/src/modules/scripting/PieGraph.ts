/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { PrivateWorker, ScriptingApi } from "./console.worker-types";
import {
    IPieGraphLayout, IPieLayoutData, IPieGraphDataPoint, IPieDemoData, IResultSetRow,
} from "../../components/ResultView/graphs/PieGraphImpl";

// eslint-disable-next-line no-restricted-globals
const ctx: PrivateWorker = self as unknown as PrivateWorker;

// This is the intermediate Pie graph class, which can be used by the user in JS/TS editors.
// It forwards all Pie graph calls to the PieGraphImpl component in the application.
export class PieGraphProxy {
    public static readonly layout: IPieLayoutData = {
        mediumPie: {
            width: 700,
            height: 300,
            innerRadius: 0,
            outerRadius: 120,
            centerX: 300,
            centerY: 150,
        },
        mediumDonut: {
            width: 700,
            height: 240,
            innerRadius: 60,
            outerRadius: 100,
            centerX: 350,
            centerY: 120,
        },
        largePie: {
            width: 700,
            height: 500,
            innerRadius: 0,
            outerRadius: 150,
            centerX: 500,
            centerY: 250,
        },
        largeDonut: {
            width: 700,
            height: 450,
            innerRadius: 130,
            outerRadius: 200,
            centerX: 500,
            centerY: 250,
        },

    };

    public static readonly demoData: IPieDemoData = {
        budget: [
            { label: "Third Party Logistics and Packaging", value: 29.2, color: "#004a83" },
            { label: "General Retail and Wholesale", value: 20.8, color: "#3c74a5" },
            { label: "Manufacturing", value: 16.8, color: "#4281b6" },
            { label: "E-Commerce", value: 10.7, color: "#4789c3" },
            { label: "Food and Beverage", value: 9.5, color: "#4b93ce" },
            { label: "Constructions/Improvements/Repair", value: 4.5, color: "#65a1d7" },
            { label: "Furniture and Appliances", value: 4.1, color: "#84b0dd" },
            { label: "Motor Vehicles/Tires/Parts", value: 3.7, color: "#9cbee0" },
        ] as IPieGraphDataPoint[],
        spending: [
            { label: "Federal Contracts & Grants", value: 19, color: "#5588b7" },
            { label: "State Support", value: 27, color: "#ba5b55" },
            { label: "Tuition & Fees", value: 18, color: "#9bb260" },
            { label: "Educational Activities & Auxiliaries", value: 11, color: "#866ea0" },
            { label: "Other C&Gs", value: 11, color: "#51a7be" },
            { label: "Private Gifts", value: 8, color: "#e4964e" },
            { label: "Other", value: 6, color: "#a2b9d9" },
        ] as IPieGraphDataPoint[],
    };

    protected contextId: string | undefined;
    protected taskId: number | undefined;

    public constructor(layout: IPieGraphLayout, data?: IPieGraphDataPoint[] | IResultSetRow[]) {
        // Store the current context id so that we always send additional data points to the same output area,
        // even if that is triggered asynchronously.
        this.contextId = ctx.currentContext;
        this.taskId = ctx.currentTaskId;

        ctx.postMessage({
            taskId: this.taskId,
            data: {
                api: ScriptingApi.PieGraphCreate,
                graphLayout: layout,
                graphData: data,
                contextId: this.contextId,
                final: true,
            },
        });
    }

    public addDataPoints(data: IPieGraphDataPoint[]): void {
        ctx.postMessage({
            taskId: this.taskId,
            data: {
                api: ScriptingApi.PieGraphAddPoints,
                graphData: data,
                contextId: this.contextId,
            },
        });
    }

}
