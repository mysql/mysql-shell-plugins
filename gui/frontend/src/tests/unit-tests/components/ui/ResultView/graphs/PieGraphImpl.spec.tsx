/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mount } from "enzyme";
import React from "react";

import { PieGraphImpl } from "../../../../../../components/ResultView";
import { PieGraphProxy } from "../../../../../../modules/scripting/PieGraph";
import { snapshotFromWrapper } from "../../../../test-helpers";

describe("Pie Graph Impl Tests", (): void => {

    it("Standard Rendering (no data)", () => {
        const component = mount(
            <PieGraphImpl />,
        );
        expect(snapshotFromWrapper(component)).toMatchSnapshot();

        component.unmount();
    });

    it("Rendering (invalid data)", () => {
        const datum = { value: 3.7 };
        const data = new Array(150).fill(datum);
        const component = mount(
            <PieGraphImpl
                width={-100}
                height={100 / 0}
                centerX={100 / 0}
                centerY={"abc" as unknown as number}
                innerRadius={200}
                outerRadius={-200}
                pointData={data}
            />,
        );
        expect(snapshotFromWrapper(component)).toMatchSnapshot();

        component.unmount();
    });

    it("Render With Data", () => {
        const component = mount(
            <PieGraphImpl
                width={100}
                height={100}
                innerRadius={100 / 0}
                pointData={[{ value: 1, color: "#123" }, { value: 2 }]}
            />,
        );
        expect(snapshotFromWrapper(component)).toMatchSnapshot();

        component.setProps({ width: 200, pointData: PieGraphProxy.demoData.budget });

        component.unmount();
    });

});
