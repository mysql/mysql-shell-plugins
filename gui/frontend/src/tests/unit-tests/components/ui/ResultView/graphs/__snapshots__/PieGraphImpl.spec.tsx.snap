// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pie Graph Impl Tests Render With Data 1`] = `
<PieGraphImpl
  height={100}
  innerRadius={Infinity}
  pointData={
    Array [
      Object {
        "color": "#123",
        "value": 1,
      },
      Object {
        "value": 2,
      },
    ]
  }
  width={100}
>
  <svg
    className="msg pieChart"
  />
</PieGraphImpl>
`;

exports[`Pie Graph Impl Tests Rendering (invalid data) 1`] = `
<PieGraphImpl
  centerX={Infinity}
  centerY="abc"
  height={Infinity}
  innerRadius={200}
  outerRadius={-200}
  pointData={
    Array [
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
      Object {
        "value": 3.7,
      },
    ]
  }
  width={-100}
>
  <svg
    className="msg pieChart"
  />
</PieGraphImpl>
`;

exports[`Pie Graph Impl Tests Standard Rendering (no data) 1`] = `
<PieGraphImpl>
  <svg
    className="msg pieChart"
  />
</PieGraphImpl>
`;
